// @tryvital/vital-link/src/types/index.ts
import React from 'react';

interface CommonVitalLinkOptions {
  // A function that is called when a user has successfully connecter an Item.
  // The function should expect two arguments, the public_key and a metadata object
  onSuccess: Function;
  // A callback that is called when a user has specifically exited Link flow
  onExit?: Function;
  // A callback that is called when a user has encountered an Error
  onError?: Function;
  // A callback that is called when the Link module has finished loading.
  // Calls to vitalLinkHandler.open() prior to the onLoad callback will be
  // delayed until the module is fully loaded.
  onLoad?: Function;

  env: string;
}

export type VitalLinkOptionsWithLinkToken = CommonVitalLinkOptions & {};

// Either the publicKey or the token field must be configured. The publicKey
// is deprecated so prefer to initialize Link with a Link Token instead.
export type VitalLinkOptions = VitalLinkOptionsWithLinkToken;

export type VitalLinkPropTypes = VitalLinkOptions & {
  children: React.ReactNode;
  className?: string;
  style?: React.CSSProperties;
};

export interface Vital {
  open: Function;
  exit: Function;
  create: Function;
  destroy: Function;
}

declare global {
  interface Window {
    Vital: Vital;
  }
}
